### PR TITLE
Keepalive

### PR DESCRIPTION
**Pull request by akashihi**
# 2056
https://github.com/Project-OSRM/osrm-backend/issues/2056

A keepalive will be a nice addition to the current http server.

### CODE REVIEW

File: src/server/request_parser.cpp [hunk 1]
```diff
@@ -7,6 +7,7 @@
 #include <boost/algorithm/string/predicate.hpp>
 
 #include <string>
+#include <util/log.hpp>
```
Comment (gardster) [inline src/server/request_parser.cpp]: Why do you need this header?

Comment (akashihi) [reply src/server/request_parser.cpp]: Development leftovers, fixed

**Comment by akashihi** [comment]
> Please add keep_alive information and corresponding constants (5 seconds, 512 connections to the documentation)

done

**Comment by gardster** [comment]
Code looks good to me.
Please add the corresponding changelog entry.
Also there are several commits about node flags. Do they relevant for your change?

**Comment by akashihi** [comment]
> Please add the corresponding changelog entry.
Done
> Also there are several commits about node flags. Do they relevant for your change?
Rebase glitch, should be fixed now

**Comment by hy05190134** [comment]
void Connection::start()                                                                                                     
  29 {                                                                                                                            
  30     TCP_socket.async_read_some(                                                                                              
  31         boost::asio::buffer(incoming_data_buffer),                                                                           
  32         strand.wrap(boost::bind(&Connection::handle_read,                                                                    
  33                                 this->shared_from_this(),                                                                    
  34                                 boost::asio::placeholders::error,                                                            
  35                                 boost::asio::placeholders::bytes_transferred)));                                             
  36                                                                                                                              
  37     if (keep_alive)                                                                                                          
  38     {                                                                                                                        
  39         // Ok, we know it is not a first request, as we switched to keepalive                                                
  40         timer.cancel();                                                                                                      
  41         timer.expires_from_now(boost::posix_time::seconds(keepalive_timeout));                                               
  42         timer.async_wait(std::bind(                                                                                          
  43             &Connection::handle_timeout, this->shared_from_this(), std::placeholders::_1));                                  
  44     }                                                                                                                        
  45 }                                                                                                                            
  46                                                                                                                              
  47 void Connection::handle_read(const boost::system::error_code &error, std::size_t bytes_transferred)                          
  48 {                                                                                                                            
  49     if (error)                                                                                                               
  50     {                                                                                                                        
  51         return;                                                                                                              
  52     }                                                                                                                        
  53                                                                                                                              
  54     if (keep_alive)                                                                                                          
  55     {                                                                                                                        
  56         timer.cancel();                                                                                                      
  57         timer.expires_from_now(boost::posix_time::seconds(0));                                                               
  58     }  

I wonder the keepalive whether work since you set expire_from_now in start() then expires_from_now in handle_read(), but every start() will execute handle_read()

**Comment by hy05190134** [comment]
void Connection::handle_read(const boost::system::error_code &error, std::size_t bytes_transferred)                          
  48 {                                                                                                                            
  49     if (error)                                                                                                               
  50     {                                                                                                                        
  51         return;                                                                                                              
  52     }                                                                                                                        
  53                

when read some bytes error, why not execute handle_shutdown right now?

**Comment by mjjbell** [comment]
> when read some bytes error, why not execute handle_shutdown right now?

This was added recently to fix a file exhaustion issue - see #6112

**Pull request merged**